mdevctl

```
@@ -314,6 +314,153 @@ unique_uuid() {
     echo "$uuid"
 }
 
+callout_script="$persist_base/callouts/command.d"
+callout_notification="$persist_base/notification/notifiers.d"
+
+state="none"
+
+dir_is_empty() {
+    path="$1"
+
+    if [ -n "$(ls -A "$path")" ]; then
+        return 1
+    fi
+
+    return 0
+}
+
+get_callout_script() {
+    path="$1"
+
+    if [ "$script" != "" ]; then
+        return 0
+    else
+        for s in $path/*; do
+            script=$($s -t "$type")
+            if [ "$script" != "" ]; then
+                return 0
+            fi
+        done
+    fi
+
+    return 1
+}
+
+# Not all commands provide all necessary info for the callouts 
+# Some sysfs devices need some extra work
+get_device_parameters() {
+    uuid="$1"
+    parent="$2"
+    type="$3"
+
+    if [ -z "$parent" ]; then
+        count=$(find "$persist_base" -name "$uuid" -type f | wc -l)
```

This 'count' variable is really confusing; in the end, we are checking for existence or non-exisitence, and ambiguity would be an error?

---

How does that work for 'undefine'? We allow ambiguity there (configs for a uuid are removed for all parents).

---

In Bash, all variables by default are defined as global, even if declared inside the function. $config is also a global variable declared at line 20. Once $uuid, $config, $parent and $type variables are populated, they are still accessible even if the config file is removed.

Before the initial callout of 'undefine', $uuid is provided by the user. Line 356 to 371 is going to populate $parent based on the $uuid. If no error occurs, line 384 will populate $config by reading into the existing config file of the $uuid and $parent. Lastly, line 389 is going retrieve the $type.

---

Yes, but don't we fail to populate parent if we have multiple configs?

---

That's correct. get_device_parameter function will return 1 if we have multiple configs. $parent is not populated, which means it's an empty string. Later, callout script from the initial callout should detect the invalid $parent, and output any error.

---

But what does that mean for undefine, where we do allow ambuguity? We don't want to return an error in that case, and I don't see the callouts invoked for each occurence. What am I missing? (Also, isn't the way callout() evaluates positional parameters a problem?)

---

Okay, I understand what you mean now. Sorry that I didn't pay attention to the ambiguity of undefine.

I took some time and tested undefine multiple configs with the same uuid. I realized that the callout script didn’t invoke at all for this situation.

The get_device_parameters function returns 1 when it detects there are multiple configs exist. Therefore, $type variable is not populated. Then the callout function returns 0 because the get_callout_script function didn’t locate any script if no type was provided. In the end, all configs are still being removed, but without invoking any callouts.

We had some discussion on this topic. I think undefine is the only case that allows ambiguity (correct me if I'm wrong). What we plan to do for undefine if multiple configs have the same uuid is to invoke callouts for an individual device, and remove one device at a time. Exit the program until all the configs with the same uuid are checked.

Please let me know if you have any suggestions for this. Thank you!

---

Invoking the callouts for any of the uuid/parent combinations looks like the right thing to do.

As for implementation, I'm not sure. You'd probably need to loop both before and after removing the definitions, but any caching of values would probably be ugly.